### DISCUSSION TITLE
Issue: Fatal Python error: Cannot recover from stack overflow.

### DISCUSSION THREAD

**Discussion by imadmoussa1** [fastapi/fastapi#7235] [Questions] [answered]
### First check

* [ ] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorials in the docs and didn't find an answer.
* [ ] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [ ] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [ ] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [ ] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask 

### Description

I have this issue on my QA server, but I was not able to replicate it locally.
My code based on the latest version of https://github.com/tiangolo/full-stack-fastapi-postgresql using the base class in CRUD,
this error happened on the update function

```
Fatal Python error: Cannot recover from stack overflow.

Thread 0x00007f615d0c0700 (most recent call first):
  File "/usr/local/lib/python3.7/concurrent/futures/thread.py", line 78 in _worker
  File "/usr/local/lib/python3.7/threading.py", line 870 in run
  File "/usr/local/lib/python3.7/threading.py", line 926 in _bootstrap_inner
  File "/usr/local/lib/python3.7/threading.py", line 890 in _bootstrap

Thread 0x00007f615d8c1700 (most recent call first):
  File "/usr/local/lib/python3.7/concurrent/futures/thread.py", line 78 in _worker
  File "/usr/local/lib/python3.7/threading.py", line 870 in run
  File "/usr/local/lib/python3.7/threading.py", line 926 in _bootstrap_inner
  File "/usr/local/lib/python3.7/threading.py", line 890 in _bootstrap

Current thread 0x00007f615e19f700 (most recent call first):
  File "<string>", line 1 in <lambda>
  File "/usr/local/lib/python3.7/site-packages/sqlalchemy/sql/operators.py", line 432 in __getitem__
```

### Environment

* OS: [e.g. Linux / Windows / macOS]: ubuntu
* FastAPI Version [e.g. 0.3.0]:  0.61.0
* Python version: 3.7

### Additional context

```
  def update(self, db: Session, *, db_obj: ModelType, obj_in: Union[UpdateSchemaType, Dict[str, Any]]) -> ModelType:
    obj_data = jsonable_encoder(db_obj)
    if isinstance(obj_in, dict):
      update_data = obj_in
    else:
      update_data = obj_in.dict(exclude_unset=True)
    for field in obj_data:
      if field in update_data:
        setattr(db_obj, field, update_data[field])
    db.add(db_obj)
    db.commit()
    db.refresh(db_obj)
    return db_obj
```

**Comment by Kludex** [comment]
This was caused by the [new release (0.61.0)](https://fastapi.tiangolo.com/release-notes/#breaking-changes) (jsonable_encoder).

The problem is the line that contains: `obj_data = jsonable_encoder(db_obj)`

Possible solution:
```python
# Your function
  def update(self, db: Session, *, db_obj: ModelType, obj_in: Union[UpdateSchemaType, Dict[str, Any]]) -> ModelType:
    obj_data = db_obj.as_dict() # HERE!
    if isinstance(obj_in, dict):
      update_data = obj_in
    else:
      update_data = obj_in.dict(exclude_unset=True)
    for field in obj_data:
      if field in update_data:
        setattr(db_obj, field, update_data[field])
    db.add(db_obj)
    db.commit()
    db.refresh(db_obj)
    return db_obj

# SQLAlchemy Base object
@as_declarative()
class Base:
    def as_dict(self) -> dict:
        return {c.key: getattr(self, c.key) for c in inspect(self).mapper.column_attrs}
```

But I don't know why we have this error message, I need to check. 

EDIT: As reminded by @imadmoussa1:
```python 
from sqlalchemy import inspect
```

**Comment by anh-collab** [comment]
I have the same FATAL errors when running normal select statement with `databases/encode`.
```
async def list_creators(
    skip: int = 0,
    limit: int = 0,
    manager_id: Optional[int] = None,
    mapping_group_id: Optional[int] = None,
):
    # basic select statement
    base_query = models.Creator.__table__.select()
    # if input manager_id
    # add where condition of mapping_groups with manager_id
    if manager_id:
        base_query = base_query.where(
            models.Creator.mapping_group_id.in_(
                select([column("mapping_group_id")])
                .select_from(models.MappingGroupMember.__table__)
                .where(models.MappingGroupMember.manager_id == manager_id)
            )
        )
    # if mapping_group_id
    # add where condition of mapping_groups with mapping_group_id
    if mapping_group_id:
        base_query = base_query.where(
            models.Creator.mapping_group_id == mapping_group_id
        )
    # add skip and limit
    query = base_query.offset(skip).limit(limit) if limit else base_query.offset(skip)
    return await database.fetch_all(query)
``` 
```
@router.get(
    "/list",
)
async def read_creators(
    skip: int = 0,
    limit: int = 100,
    manager_id: Optional[int] = None,
    mapping_group_id: Optional[int] = None,
):
    if manager_id and mapping_group_id:
        raise HTTPException(
            status_code=400,
            detail="Please do not provide manager_id and mapping_group_id at the same time",
        )
    return await crud.list_creators(
        skip=skip, limit=limit, manager_id=manager_id, mapping_group_id=mapping_group_id
    )
```
Any help would be appreciated. This did not happen before.

**Comment by duz-sg** [comment]
Same issue here, this crash does not give any trace info, took some time to trace the offending code, as outlined earlier:
https://github.com/tiangolo/full-stack-fastapi-postgresql/blob/master/%7B%7Bcookiecutter.project_slug%7D%7D/backend/app/app/crud/base.py#L49

**Comment by imadmoussa1** [comment]
@Kludex Thank you, Your solution fixed my problem.

**Comment by Kludex** [comment]
You're welcome! @imadmoussa1 Do you mind closing the issue then? :sunglasses: :tada:

**Comment by sv8083** [comment]
> ```python
> def as_dict(self) -> dict:
>         return {c.key: getattr(self, c.key) for c in inspect(self).mapper.column_attrs}
> ```

Hi @Kludex 
I applied the fix, but it throws me below error:
```
  File ".\crud\base.py", line 49, in update
    obj_data = db_obj.as_dict()
  File ".\db\base_class.py", line 15, in as_dict
    return {c.key: getattr(self, c.key) for c in inspect(self).mapper.column_attrs}
TypeError: 'module' object is not callable
```

**Comment by imadmoussa1** [comment]
Hi @sv8083 you should import the inspect from sqlalchemy
 
`from sqlalchemy import inspect`

**Comment by plankter** [comment]
Does it mean jsonable_encoder is broken? It is unclear why the issue was closed.

**Comment by imadmoussa1** [comment]
@plankter jsonable_encoder is not broken, I am using it in different functions, but on update function we had errors. fixed by replacing the jsonable_encoder with a custom function.

**Comment by duz-sg** [comment]
This fix does not help in my case, after applying the above fix, I still get the error message:

```shell
Fatal Python error: Cannot recover from stack overflow.

Thread 0x00007fa28e7bc700 (most recent call first):
  File "/usr/local/lib/python3.7/concurrent/futures/thread.py", line 78 in _worker
  File "/usr/local/lib/python3.7/threading.py", line 870 in run
  File "/usr/local/lib/python3.7/threading.py", line 926 in _bootstrap_inner
  File "/usr/local/lib/python3.7/threading.py", line 890 in _bootstrap

Thread 0x00007fa28effd700 (most recent call first):
  File "/usr/local/lib/python3.7/concurrent/futures/thread.py", line 78 in _worker
  File "/usr/local/lib/python3.7/threading.py", line 870 in run
  File "/usr/local/lib/python3.7/threading.py", line 926 in _bootstrap_inner
  File "/usr/local/lib/python3.7/threading.py", line 890 in _bootstrap

Current thread 0x00007fa28f7fe700 (most recent call first):
  File "/usr/local/lib/python3.7/site-packages/sqlalchemy/sql/type_api.py", line 67 in operate
  File "<string>", line 1 in <lambda>
  File "/usr/local/lib/python3.7/site-packages/sqlalchemy/sql/operators.py", line 432 in __getitem__
  File "/usr/local/lib/python3.7/site-packages/sqlalchemy/sql/elements.py", line 762 in operate
```

Update, fixed my issue:
```python
    check_item = crud.item.get_by_user(
        db=db, id=item_id, user_id=current_user.id
    )
    pd_check_item = ItemCheck.from_orm(check_item) # Fix: convert from DB model to pydantic model 
    check_data = jsonable_encoder(pd_check_item, exclude_none=True) # Breaks if accept db model directly
    return check_data
```

It seems `jsonable_encoder` can no longer accept a DB model and convert the data, it only accepts a pydantic model.

**Comment by shawnwall** [comment]
Another approach I was trying locally was to use load a pydantic model from the orm model, get the dict(), then iterate that, such as:

`obj_data = UserInDB.from_orm(db_obj).dict()`

...now I know that doesnt perfectly fit into `CRUDBase` unless we refactor it to also hold onto a `DBSchemaType` in addition to `CreateSchemaType` and `UpdateSchemaType`

**Comment by l1725821492** [comment]
> This fix does not help in my case, after applying the above fix, I still get the error message:
> 
> ```shell
> Fatal Python error: Cannot recover from stack overflow.
> 
> Thread 0x00007fa28e7bc700 (most recent call first):
>   File "/usr/local/lib/python3.7/concurrent/futures/thread.py", line 78 in _worker
>   File "/usr/local/lib/python3.7/threading.py", line 870 in run
>   File "/usr/local/lib/python3.7/threading.py", line 926 in _bootstrap_inner
>   File "/usr/local/lib/python3.7/threading.py", line 890 in _bootstrap
> 
> Thread 0x00007fa28effd700 (most recent call first):
>   File "/usr/local/lib/python3.7/concurrent/futures/thread.py", line 78 in _worker
>   File "/usr/local/lib/python3.7/threading.py", line 870 in run
>   File "/usr/local/lib/python3.7/threading.py", line 926 in _bootstrap_inner
>   File "/usr/local/lib/python3.7/threading.py", line 890 in _bootstrap
> 
> Current thread 0x00007fa28f7fe700 (most recent call first):
>   File "/usr/local/lib/python3.7/site-packages/sqlalchemy/sql/type_api.py", line 67 in operate
>   File "<string>", line 1 in <lambda>
>   File "/usr/local/lib/python3.7/site-packages/sqlalchemy/sql/operators.py", line 432 in __getitem__
>   File "/usr/local/lib/python3.7/site-packages/sqlalchemy/sql/elements.py", line 762 in operate
> ```
> 
> Update, fixed my issue:
> 
> ```python
>     check_item = crud.item.get_by_user(
>         db=db, id=item_id, user_id=current_user.id
>     )
>     pd_check_item = ItemCheck.from_orm(check_item) # Fix: convert from DB model to pydantic model 
>     check_data = jsonable_encoder(pd_check_item, exclude_none=True) # Breaks if accept db model directly
>     return check_data
> ```
> 
> It seems `jsonable_encoder` can no longer accept a DB model and convert the data, it only accepts a pydantic model.

> This fix does not help in my case, after applying the above fix, I still get the error message:
> 
> ```shell
> Fatal Python error: Cannot recover from stack overflow.
> 
> Thread 0x00007fa28e7bc700 (most recent call first):
>   File "/usr/local/lib/python3.7/concurrent/futures/thread.py", line 78 in _worker
>   File "/usr/local/lib/python3.7/threading.py", line 870 in run
>   File "/usr/local/lib/python3.7/threading.py", line 926 in _bootstrap_inner
>   File "/usr/local/lib/python3.7/threading.py", line 890 in _bootstrap
> 
> Thread 0x00007fa28effd700 (most recent call first):
>   File "/usr/local/lib/python3.7/concurrent/futures/thread.py", line 78 in _worker
>   File "/usr/local/lib/python3.7/threading.py", line 870 in run
>   File "/usr/local/lib/python3.7/threading.py", line 926 in _bootstrap_inner
>   File "/usr/local/lib/python3.7/threading.py", line 890 in _bootstrap
> 
> Current thread 0x00007fa28f7fe700 (most recent call first):
>   File "/usr/local/lib/python3.7/site-packages/sqlalchemy/sql/type_api.py", line 67 in operate
>   File "<string>", line 1 in <lambda>
>   File "/usr/local/lib/python3.7/site-packages/sqlalchemy/sql/operators.py", line 432 in __getitem__
>   File "/usr/local/lib/python3.7/site-packages/sqlalchemy/sql/elements.py", line 762 in operate
> ```
> 
> Update, fixed my issue:
> 
> ```python
>     check_item = crud.item.get_by_user(
>         db=db, id=item_id, user_id=current_user.id
>     )
>     pd_check_item = ItemCheck.from_orm(check_item) # Fix: convert from DB model to pydantic model 
>     check_data = jsonable_encoder(pd_check_item, exclude_none=True) # Breaks if accept db model directly
>     return check_data
> ```
> 
> It seems `jsonable_encoder` can no longer accept a DB model and convert the data, it only accepts a pydantic model.

How to import ItemCheck

**Comment by duz-sg** [comment]
> How to import ItemCheck

I'm using the project generator, so you can define your own pydantic models like `ItemCheck`, you can refer to the samples in project generator here:
https://github.com/tiangolo/full-stack-fastapi-postgresql/blob/master/%7B%7Bcookiecutter.project_slug%7D%7D/backend/app/app/schemas/item.py#L33
As long as it is based on `ItemInDBBase`, it will support `orm_mode`

**Comment by chrisjsimpson** [comment]
For me, and also using sqlalchemy, I was simply forgetting to add `response_model=schemaName` to my endpoint:

Before:

```
@app.get("/users/me")
async def read_users_me(current_user: schemas.User = Depends(get_current_active_user)):
    return current_user
```

After:
```
@app.get("/users/me", response_model=schemas.User)
async def read_users_me(current_user: schemas.User = Depends(get_current_active_user)):
    return current_user
```

**Comment by tiangolo** [comment, marked as answer]
Thanks for the discussion here everyone! I'm sorry for the inconveniences :facepalm: . This was fixed in `0.61.1`. The current latest is `0.62.0`.

**Comment by TuckerMogren** [comment]
I am currently getting this issue. 

Fatal Python error: _Py_CheckRecursiveCall: Cannot recover from stack overflow.
Python runtime state: initialized

Thread 0x000045ac (most recent call first):
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.9_3.9.2032.0_x64__qbz5n2kfra8p0\lib\threading.py", line 316 in wait
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.9_3.9.2032.0_x64__qbz5n2kfra8p0\lib\queue.py", line 180 in get
  File "C:\Users\TuckerMogren\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\LocalCache\local-packages\Python39\site-packages\aiosqlite\core.py", line 95 in run
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.9_3.9.2032.0_x64__qbz5n2kfra8p0\lib\threading.py", line 973 in _bootstrap_inner
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.9_3.9.2032.0_x64__qbz5n2kfra8p0\lib\threading.py", line 930 in _bootstrap

Current thread 0x00001e04 (most recent call first):
  File "C:\Program Files\WindowsApps\PythonSoftwareFoundation.Python.3.9_3.9.2032.0_x64__qbz5n2kfra8p0\lib\abc.py", line 119 in __instancecheck__
  File "C:\Users\TuckerMogren\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\LocalCache\local-packages\Python39\site-packages\fastapi\encoders.py", line 44 in jsonable_encoder


when I call the following function

@app.get("/api/v1/photo/{photoID}")
async def getPhoto(photoID, res: Response):
    try:
        statement = photos.select(photos.c.userID).where(photos.c.photoID == photoID)
        results = await database.fetch_one(query=statement)
        if results == []:
            raise AssertionError()
    except AssertionError as e:
        res.status_code = status.HTTP_404_NOT_FOUND
        return [{"Status": res.status_code}, {"ErrorMessage":"No data found in database for photoID: " + photoID}]
    except Error as e:
        res.status_code = status.HTTP_400_BAD_REQUEST
        return [{"Status": res.status_code}, {"ErrorMessage":e.args}]
    except:
        res.status_code = status.HTTP_500_INTERNAL_SERVER_ERROR
        return {"Message":"Unknown, fatal error"}
    res.status_code = status.HTTP_200_OK
    return [{"Status": res.status_code}, {"PhotoID":photoID}, {"Photo":photos.c.photo} ]

**Comment by 1067511899** [comment]
I get the same issue.

I do a little research:
newresult = db.query(DonateOrder.donorName, PayTrade.totalAmount, PayTrade.paymentTime).filter(
        DonateOrder.payTradeId == PayTrade.id).filter(DonateOrder.projectId == id).filter(
        DonateOrder.orderStatus == 'success').filter(PayTrade.paymentTime <= result).limit(10)

    print(type(newresult))  # <class 'sqlalchemy.orm.query.Query'>
    return newresult  # Fatal Python error as list below
    return [row._asdict() for row in newresult]  # works fine

btw:
SQLAlchemy                         1.4.22
fastapi                            0.72.0
Python 3.9.7 (default, Sep 16 2021, 16:59:28) [MSC v.1916 64 bit (AMD64)] :: Anaconda, Inc. on win32


Fatal Python error: _Py_CheckRecursiveCall: Cannot recover from stack overflow.
Python runtime state: initialized

Thread 0x00002824 (most recent call first):
  File "G:\ProgramData\Anaconda3\lib\threading.py", line 312 in wait
  File "G:\ProgramData\Anaconda3\lib\queue.py", line 171 in get
  File "G:\ProgramData\Anaconda3\lib\site-packages\anyio\_backends\_asyncio.py", line 744 in run
  File "G:\ProgramData\Anaconda3\lib\threading.py", line 973 in _bootstrap_inner
  File "G:\ProgramData\Anaconda3\lib\threading.py", line 930 in _bootstrap

Current thread 0x00001dd8 (most recent call first):
  File "G:\ProgramData\Anaconda3\lib\site-packages\sqlalchemy\sql\operators.py", line 434 in __getitem__
  File "G:\ProgramData\Anaconda3\lib\site-packages\sqlalchemy\sql\elements.py", line 865 in operate
  File "G:\ProgramData\Anaconda3\lib\site-packages\sqlalchemy\sql\operators.py", line 434 in __getitem__
  File "G:\ProgramData\Anaconda3\lib\site-packages\sqlalchemy\orm\properties.py", line 430 in operate
  File "G:\ProgramData\Anaconda3\lib\site-packages\sqlalchemy\sql\operators.py", line 434 in __getitem__
  File "G:\ProgramData\Anaconda3\lib\site-packages\sqlalchemy\orm\attributes.py", line 317 in operate
  File "G:\ProgramData\Anaconda3\lib\site-packages\sqlalchemy\sql\operators.py", line 434 in __getitem__
  ............

**Comment by seano314** [comment]
Is there a recommended way to do this with Strawberry that doesn't depend on Pydantic?

If I'm reading through `jsonable_encoder` correctly it looks like the base case to stop the recursion is when the base Pydantic model is hit via [isinstance(obj, BaseModel)](https://github.com/tiangolo/fastapi/blob/master/fastapi/encoders.py#L52)? Additionally the suggested code above depends on `update_data = obj_in.dict(exclude_unset=True)`, which is also [Pydantic based](https://pydantic-docs.helpmanual.io/usage/exporting_models/#modeldict).

If I'm reading through the Strawberry docs correctly, setting the schema as Pydantic is possible but is still an [experimental feature](https://strawberry.rocks/docs/integrations/pydantic). 

Is there another solution that can fix this problem that doesn't involve converting Strawberry to Pydantic? 

(Note I'm asking this here instead of Strawberry as it is the recommended framework as per [FastAPI documentation](https://fastapi.tiangolo.com/fr/advanced/graphql/#graphql-with-strawberry))

**Comment by tiangolo** [comment]
If you are having problems with this, or similar, please create a new issue, following the template guide, and creating a minimal self contained example that replicates it. In many cases you might see the same error, but actually have different cause, and the only way to find it is by following that properly and writing the example code to reproduce it.
